dns_sd_bonjour.c

```diff
@@ -44,124 +52,197 @@ static void __cfnet_browser_cb (
 		DEBUG("DNS SD: No more entries coming.\n");
-		goto stop_browsing;
+		CFNetServiceBrowserStopSearch(browser, &anError);
 	}
-	
+
+exit:
+	iio_mutex_unlock(dd->lock);
 	return;
 
 stop_browsing:
-	
 	CFNetServiceBrowserStopSearch(browser, &anError);
 }
 
-int discover_host(char *addr_str, size_t addr_len, uint16_t *port)
+int dnssd_find_hosts(struct dns_sd_discovery_data ** ddata)
 {
 	int ret = 0;
-	
-	struct cfnet_discovery_data ddata = { FALSE, FALSE };
-	CFNetServiceClientContext clientContext = { 0, &ddata, NULL, NULL, NULL };
-	CFStreamError error;
-	Boolean result;
+	struct dns_sd_discovery_data *d;
 
-	CFStringRef type = CFSTR("_iio._tcp.");
-	CFStringRef domain = CFSTR("");
+	DEBUG("DNS SD: Start service discovery.\n");
 
-	DEBUG("DNS SD: Resolving hostname.");
+	if (new_discovery_data(&d) < 0) {
+		return -ENOMEM;
+	}
+
+	d->lock = iio_mutex_create();
+	if (!d->lock) {
+		dnssd_free_all_discovery_data(d);
+		return -ENOMEM;
+	}
 
-	CFNetServiceBrowserRef gServiceBrowserRef = CFNetServiceBrowserCreate(
+	CFNetServiceClientContext clientContext = { 0, d, NULL, NULL, NULL };
+	CFNetServiceBrowserRef serviceBrowser = CFNetServiceBrowserCreate(
 		kCFAllocatorDefault, __cfnet_browser_cb, &clientContext);
 
-	if (gServiceBrowserRef == NULL) {
+	if (serviceBrowser == NULL) {
 		ERROR("DNS SD: Failed to create service browser.\n");
-		ret = ENOMEM;
+		dnssd_free_all_discovery_data(d);
+		ret = -ENOMEM;
 		goto exit;
 	}
-	result = CFNetServiceBrowserSearchForServices(
-		gServiceBrowserRef, domain, type, &error);
+
+	CFRunLoopRef runLoop = CFRunLoopGetCurrent();
+	CFNetServiceBrowserScheduleWithRunLoop(serviceBrowser, runLoop, kCFRunLoopDefaultMode);
+
+	CFStringRef type = CFSTR("_iio._tcp.");
+	CFStringRef domain = CFSTR("local.");
+	CFStreamError error;
+	Boolean result = CFNetServiceBrowserSearchForServices(serviceBrowser, domain, type, &error);
 
 	if (result == false) {
 		ERROR("DNS SD: CFNetServiceBrowserSearchForServices failed (domain = %ld, error = %d)\n",
 			(long)error.domain, error.error);
-		ret = ENXIO;
-		goto free_browser;
-	}
-	
-	// Resolved address and port. ipv4 has precedence over ipv6.
-	*port = ddata.port;
-	if (ddata.have_v4) {
-		strncpy(addr_str, ddata.address_v4, addr_len);
-	}
-	else if (ddata.have_v6) {
-		strncpy(addr_str, ddata.address_v6, addr_len);
-	}
-	else {
-		ERROR("DNS SD: IIO service not found.\n");
-		ret = ENXIO;
+
+		ret = -ENXIO;
+	} else {
+		CFRunLoopRunResult runRes = CFRunLoopRunInMode(kCFRunLoopDefaultMode, 2, TRUE);
+
+		if (runRes != kCFRunLoopRunHandledSource) {
+			if (runRes == kCFRunLoopRunFinished)
+				ERROR("DSN SD: CFRunLoopRunInMode completed kCFRunLoopRunFinished (%d)\n", runRes);
+			else if (runRes == kCFRunLoopRunStopped)
+				ERROR("DSN SD: CFRunLoopRunInMode completed kCFRunLoopRunStopped (%d)\n", runRes);
+			else if (runRes == kCFRunLoopRunTimedOut)
+				ERROR("DSN SD: CFRunLoopRunInMode completed kCFRunLoopRunTimedOut (%d)\n", runRes);
```

Comment:
IMHO timeout is not an error but indication that no iio services were found.
```
$ iio.framework/Tools/iio_info -s
Library version: 0.19 (git tag: 478b1ac)
Compiled with backends: xml ip
ERROR: DSN SD: CFRunLoopRunInMode completed kCFRunLoopRunTimedOut (3)
No contexts found.
```

Reply:
fixed